xtask/src/containers.rs

```diff
@@ -0,0 +1,159 @@
+//
+// Copyright 2023 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+use crate::{
+    internal::{Cmd, Runnable, Step},
+    workspace_path,
+};
+
+fn build_kernel() -> Step {
+    Step::Single {
+        name: "build kernel".to_string(),
+        command: Cmd::new("just", vec!["oak_containers_kernel"]),
+    }
+}
+
+fn build_system_image() -> Step {
+    Step::Single {
+        name: "build system image".to_string(),
+        command: Cmd::new("just", vec!["oak_containers_system_image"]),
+    }
+}
+
+fn build_oak_functions_bundle() -> Step {
+    Step::Single {
+        name: "build Oak Functions bundle".to_string(),
+        command: Cmd::new(
+            "just",
+            vec!["oak_functions_containers_container_bundle_tar"],
+        ),
+    }
+}
+
+fn build_oak_functions_launcher() -> Step {
+    Step::Single {
+        name: "build Oak Functions bundle".to_string(),
+        command: Cmd::new("just", vec!["oak_functions_containers_launcher"]),
+    }
+}
+
+async fn build_prerequisites() {
+    // `just` steps can clash with each other, so ensure that we're not trying to build things like
+    // the system image concurrently.
+    static BUILD: tokio::sync::OnceCell<()> = tokio::sync::OnceCell::const_new();
+    BUILD
+        .get_or_init(async || {
+            tokio::join!(
+                crate::testing::run_step(crate::launcher::build_stage0()),
+                crate::testing::run_step(build_kernel()),
+                crate::testing::run_step(build_system_image()),
+                crate::testing::run_step(build_oak_functions_bundle()),
+                crate::testing::run_step(build_oak_functions_launcher())
+            );
+        })
+        .await;
+}
+
+pub fn run_oak_functions_launcher_example_with_lookup_data(
```

Comment:
I wonder if we should move some of this dependency management in `just`, now that we are more familiar with it. We can still spawn out to call `just` from Rust tests, but at least we can declare the dependency tree in the `justfile` itself. Anyways, not sure it's worth doing at the moment, but we can chat about it separately.

Reply:
Would this involve using bazel for rust? Has this idea been entertained by Oak before?

Reply:
We have thought about it to the extent of "someone should play around with it to see if it is feasible", but it hasn't ever been a priority for us.